### ISSUE TITLE
Compiler caused error: Missing mandatory field Value.value

### ISSUE THREAD

**Issue by joneff** [sass/embedded-host-node#112]
First of all, bravo, for all the hard work! I am working with especially large sass project, so sass-embedded brings significant speed increase over dart-sass.

I am using a sort of hack to get all sass variables, along with their type and value like so:

```scss
// all.scss
$var1: 1;
$var2: 2px;

@import "PATH_TO_NODE_MODULES/bootstrap/scss/bootstrap.scss";
```

```scss
// variables.scss
@use "sass:meta";
@use "all.scss";

$_vars: meta.module-variables("all");

body {
    @each $key, $val in $_vars {
        var: resolve-var( unquote($key), meta.type-of( $val ), $val );
    }
}
```

Then I am compiling the file by passing a custom function like so:

```js
const fs = require('fs');
const path = require('path');
const sassEmbedded = require('sass-embedded');

const cwd = process.cwd();

let variablesJson = path.resolve( cwd, `dist/variables.json` );
let variablesScss = path.resolve( cwd, `dist/variables.scss` );
let content = {};

sassEmbedded.compile(variablesScss, {
    functions: {
        'resolve-var($key, $type, $value)': (args) => {
            const _key = args[0].toString();
            const _type = args[1].toString();
            const _val = args[2].toString();

            content[_key] = {
                type: _type,
                value: _val
            };

            return sassEmbedded.sassNull;
        }
    },
    logger: sassEmbedded.Logger.silent
});

fs.writeFileSync( variablesJson, JSON.stringify( content, null, 4 ) );
```

The above works with sass.

With sass embedded it throws an error:  `Compiler caused error: Missing mandatory field Value.value.` If I console log right after variable assignment, the last successful parsed variable is `$input-height-border`. So the one that's failing should be https://github.com/twbs/bootstrap/blob/main/scss/_variables.scss#L839. It's worth noting that I've tried with calc and it also fails. other functions pass. 

Clarification, the line in question resorts to `calc`, so that's the one that fails.

On a side note, if I comment out bootstrap.scss and console.log(_val), I get different return types:
* dartSass yields `1` and `2px`
* dartEmbedded yields `1no unit` and `2px`

**Comment by Awjin** [comment]
This looks like it might be a bug, but it is hard to tell from your code examples.

Could you provide a bare-minimum set of code that reproduces this bug?

**Comment by joneff** [comment]
I've attached a very barebones example. in the current state it will work. Uncomment `var-b` in vars.scss and things will stop working.

Also, notice how the variable is described in the resulting json:

```json
// dart
{
    "var-a": {
        "type": "number",
        "value": "3"
    }
}

//embedded
{
    "var-a": {
        "type": "number",
        "value": "3no units"
    }
}
```

If you are unwilling to download the file, you can see this gist -- https://gist.github.com/joneff/dc1d63909c8190485dc6f229710a9a71 -- though, you will need to tweak the paths a bit.

[sass-embedded-issue.zip](https://github.com/sass/embedded-host-node/files/8088422/sass-embedded-issue.zip)

**Comment by ntkme** [comment]
Two different issues:

1. `no units` in inspection, this is due to `toString` method calling `unitString` which adds `no units`
https://github.com/sass/embedded-host-node/blob/db948177ec47e0da2a421de40c8d309d48e83606/lib/src/value/number.ts#L545-L550
https://github.com/sass/embedded-host-node/blob/db948177ec47e0da2a421de40c8d309d48e83606/lib/src/value/number.ts#L682-L688
The reason that `unitString` would return new units is due to the usage in a different context to log error message requires `no unit` to be printed:
https://github.com/sass/embedded-host-node/blob/db948177ec47e0da2a421de40c8d309d48e83606/lib/src/value/number.ts#L616-L622
Fix should be simple: add a boolean flag to unitString to determine whether to output unit as empty string or `no unit`.

2. The protofy of `calc` type is support in embedded-dart-sass, but it is not implemented in embedded-host-node, to implement the feature, we will have to define the host type of `SassCalculation` first. I think the biggest challenge would be implementation equity check because the protocol does not require host to support simplification of `calc`. We can always check equality of host type by [sending them to compiler and let compiler decide](https://github.com/ntkme/sass-embedded-host-ruby/blob/340d769e8d06033b547baf0d038f2cb426046f34/spec/sass_proto_spec.rb#L6-L35). Basically, I think we still need more thoughts on how to design the host interface for `SassCalculation`.

**Comment by joneff** [comment]
@ntkme considering the PR is now merged, would that mean the fix made it in 1.51.0, that was released yesterday?

**Comment by ntkme** [comment]
@joneff Depends on which issue you are talking about. The PR only fixes the `1no unit` issue, and it is in `>=1.50.1`. The `Missing mandatory field Value.value` issue still exists and isn't likely to be fixed in short term.

**Comment by nex3** [comment]
The remaining issues is a duplicate of https://github.com/sass/sass/issues/818 (specifically the one remaining task of defining the JS API for calculation objects).
